lib/Conversion/HWToSystemC/HWToSystemC.cpp

```diff
@@ -21,6 +22,112 @@
 using namespace mlir;
 using namespace circt;
 using namespace hw;
+using namespace systemc;
+
+//===----------------------------------------------------------------------===//
+// Operation Conversion Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// This works on each HW module, creates corresponding SystemC modules, moves
+/// the body of the module into the new SystemC module by splitting up the body
+/// into field declarations, initializations done in a newly added systemc.ctor,
+/// and internal methods to be registered in the constructor.
+struct ConvertHWModule : public OpConversionPattern<HWModuleOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(HWModuleOp module, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Collect the HW module's port types.
+    FunctionType moduleType = module.getFunctionType();
+    TypeRange moduleInputs = moduleType.getInputs();
+    TypeRange moduleOutputs = moduleType.getResults();
+
+    // SystemC module port types are all expressed as block arguments to the op,
+    // so collect all of the types and add them as arguments to the SystemC
+    // module.
+    SmallVector<Type, 4> portTypes(moduleInputs);
+    portTypes.append(moduleOutputs.begin(), moduleOutputs.end());
+
+    // Collect all the port directions.
+    SmallVector<systemc::PortDirection> directions;
+    for (auto port : module.getAllPorts()) {
+      if (port.isInput())
+        directions.push_back(systemc::PortDirection::Input);
+      else if (port.isOutput())
+        directions.push_back(systemc::PortDirection::Output);
+      else
+        directions.push_back(systemc::PortDirection::InOut);
```

Comment:
Is it possible to ever exercise this code via a HW -> SystemC module conversion? Should this be an error for now?

Reply:
You can, for example, exercise it with `hw.module @someModule (%inoutArg: !hw.inout<i32>) -> () {}` which is converted to `systemc.module @someModule (sc_inout %inoutArg: !hw.inout<i32>) {}`. Should have added that to a test. We might also have to peel off the `!hw.inout` wrapper type. But I think these inout values are only really used by the SV dialect, so to error out can also be a strategy for now.